staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/controller.go

```diff
@@ -0,0 +1,551 @@
+*/
+
+package cel
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/apiserver/pkg/admission/plugin/cel/generic"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/dynamic/dynamicinformer"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+)
+
+// celAdmissionController is the top-level controller for admission control using CEL
+// it is responsible for watching policy definitions, bindings, and config param CRDs
+type celAdmissionController struct {
+	// Context under which the controller runs
+	runningContext context.Context
+
+	policyDefinitionsController generic.Controller[fantasyInternalPolicyDefinition]
+	policyBindingController     generic.Controller[fantasyInternalPolicyBinding]
+	paramsCRDControllers        map[schema.GroupVersionKind]paramInfo
+
+	// dynamicclient used to create informers to watch the param crd types
+	dynamicClient dynamic.Interface
+	restMapper    meta.RESTMapper
+
+	// Provided to the policy' sCompile function as an injected dependency to
+	// assist with compiling its expressions to CEL
+	objectConverter CELObjectConverter
+
+	// Index for each definition namespace/name, contains all binding
+	// namespace/names known to exist for that definition
+	definitionInfo map[string]definitionInfo
+
+	// Index for each bindings namespace/name. Contains compiled templates
+	// for the binding depending on the policy/param combination.
+	bindingInfos map[string]bindingInfo
+
+	// Map from namespace/name of a definition to a set of namespace/name
+	// of bindings which depend on it.
+	// All keys must have at least one dependent binding
+	// All binding names MUST exist as a key bindingInfos
+	definitionsToBindings map[string]sets.String
+
+	// Lock which protects:
+	//	- definitionInfo
+	//  - bindingInfos
+	//  - paramCRDControllers
+	//  - definitionsToBindings
+	// All other fields should be assumed constant
+	mutex sync.RWMutex
+}
+
+type definitionInfo struct {
+	// Error about the statae of the definition's configuration and the cluster
+	// preventing its enforcement or compilation.
+	// Reset every reconciliation
+	configurationError error
+
+	// Last value seen by this controller to be used in policy enforcement
+	// May be nil (due to dependentBindings possibly existing before definition)
+	lastReconciledValue fantasyInternalPolicyDefinition
+}
+
+type bindingInfo struct {
+	// Compiled CEL expressoin turned into an evaluator
+	evaluator EvaluatorFunc
+
+	// Last value seen by this controller to be used in policy enforcement
+	// May not be nil
+	lastReconciledValue fantasyInternalPolicyBinding
```

Comment:
This is in a map by-value, the code referencing this is aware that modifying it doesn't modify the copy in the map, but it still makes me nervous (about the next person to modify this code, not you). Is there a reason why you didn't make it a map to *bindingInfo?

Reply:
Mostly just personal bias. It was easy enough to change over so I've done that and checked the usages.